crates/number/src/u256_hex_or_decimal.rs

```diff
@@ -0,0 +1,80 @@
+use {
+    primitive_types::U256,
+    serde::{de, Deserializer, Serializer},
+    serde_with::{DeserializeAs, SerializeAs},
+    std::fmt,
+};
+
+pub struct HexOrDecimalU256;
+
+impl<'de> DeserializeAs<'de, U256> for HexOrDecimalU256 {
+    fn deserialize_as<D>(deserializer: D) -> Result<U256, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        deserialize(deserializer)
+    }
+}
+
+impl SerializeAs<U256> for HexOrDecimalU256 {
+    fn serialize_as<S>(source: &U256, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        super::u256_decimal::serialize(source, serializer)
+    }
+}
+
+pub fn deserialize<'de, D>(deserializer: D) -> Result<U256, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    struct Visitor {}
```

Comment:
Conceptually, `serde::de::Visitor`s are confusing, but needed because not all formats are like JSON where the type is known ahead of time (i.e. not all serialization formats can differentiate data types). One trick I've used in the past is to implement this without needing a `Visitor` is:

```rust
pub fn deserialize<'de, D>(deserializer: D) -> Result<U256, D::Error>
where
    D: Deserializer<'de>,
{
    let s = Cow::<str>::deserialize(deserializer)?;
    // ... rest of stuff
}
```

The downside (and what makes your approach better), is that the `expecting` implementation gives a more accurate message under certain error conditions.

Also, nano-nit:
```suggestion
    struct Visitor;
```